lib/active_merchant/billing/gateways/safe_charge.rb

```diff
@@ -149,26 +149,46 @@ def add_transaction_data(trans_type, post, money, options)
       end
 
       def add_payment(post, payment, options = {})
-        post[:sg_ExpMonth] = format(payment.month, :two_digits)
-        post[:sg_ExpYear] = format(payment.year, :two_digits)
-        post[:sg_CardNumber] = payment.number
-
-        if payment.is_a?(NetworkTokenizationCreditCard) && payment.source == :network_token
-          post[:sg_CAVV] = payment.payment_cryptogram
-          post[:sg_ECI] = options[:three_d_secure] && options[:three_d_secure][:eci] || '05'
-          post[:sg_IsExternalMPI] = 1
-          post[:sg_ExternalTokenProvider] = 5
-        else
-          post[:sg_CVV2] = payment.verification_value
-          post[:sg_NameOnCard] = payment.name
-          post[:sg_StoredCredentialMode] = (options[:stored_credential_mode] == true ? 1 : 0)
+        case payment
+        when String
+          add_token(post, payment)
+        when CreditCard
+          post[:sg_ExpMonth] = format(payment.month, :two_digits)
+          post[:sg_ExpYear] = format(payment.year, :two_digits)
+          post[:sg_CardNumber] = payment.number
+
+          if payment.is_a?(NetworkTokenizationCreditCard) && payment.source == :network_token
+            add_network_token(post, payment, options)
+          else
+            add_credit_card(post, payment, options)
+          end
         end
       end
 
+      def add_token(post, payment)
+        _, transaction_id, token = payment.split('|')
+
+        post[:sg_TransactionID] = transaction_id
+        post[:sg_CCToken] = token
+      end
+
+      def add_credit_card(post, payment, options)
+        post[:sg_CVV2] = payment.verification_value
+        post[:sg_NameOnCard] = payment.name
+        post[:sg_StoredCredentialMode] = (options[:stored_credential_mode] == true ? 1 : 0)
+      end
+
+      def add_network_token(post, payment, options)
+        post[:sg_CAVV] = payment.payment_cryptogram
+        post[:sg_ECI] = options[:three_d_secure] && options[:three_d_secure][:eci] || '05'
```

Comment:
Should this be the ECI for the network token or the 3DS ECI value?

Reply:
Oh that's a good question, probably the network token. Since this is not related to the ticket and it will need some additional testing I'll create a ticket for it.